packages/dashboard-core-plugins/src/panels/FileExplorerPanel.test.tsx

```diff
@@ -0,0 +1,185 @@
+import MockFileStorage from './MockFileStorage';
+
+function makeFile(basename: string, path = '/'): FileStorageItem {
+  const filename = `${path}${basename}`;
+  return {
+    basename,
+    filename,
+    type: 'file',
+    id: filename,
+  };
+}
+
+function makeFiles(count = 5) {
+  const result: FileStorageItem[] = [];
+  for (let i = 0; i < count; i += 1) {
+    result.push(makeFile(`testfile${i}`));
+  }
+  return result;
+}
+
+function makeDirectory(name: string, path = '/'): FileStorageItem {
+  const file = makeFile(name, path);
+  file.type = 'directory';
+  return file;
+}
+
+function makeDirectories(count = 5) {
+  const result: FileStorageItem[] = [];
+  for (let i = 0; i < count; i += 1) {
+    result.push(makeDirectory(`testdir${i}`));
+  }
+  return result;
+}
+
+const eventHub = {
+  emit: () => undefined,
+  on: () => undefined,
+  off: () => undefined,
+  trigger: () => undefined,
+  unbind: () => undefined,
+};
+const container: Partial<Container> = {
+  emit: () => undefined,
+  on: () => undefined,
+  off: () => undefined,
+};
+
+function makeContainer({ fileStorage }: Partial<FileExplorerPanelProps> = {}) {
+  return render(
+    <FileExplorerPanel
+      glContainer={container}
+      glEventHub={eventHub}
+      localDashboardId="TEST DASHBOARD"
+      fileStorage={fileStorage}
+      language="TEST LANGUAGE"
+      dispatch={undefined}
+    />
+  );
+}
+
+it('mounts properly and shows file list', async () => {
+  const files = makeFiles();
+
+  const fileStorage = new MockFileStorage(files);
+
+  makeContainer({ fileStorage });
+
+  const foundItems = await screen.findAllByRole('listitem');
+  expect(foundItems).toHaveLength(files.length);
+});
+
+describe('selects directory for NewItemModal correctly', () => {
+  function clickItem(itemIndex: number, options = {}): void {
+    userEvent.click(screen.getAllByRole('listitem')[itemIndex], options);
+  }
+
+  async function getNewItemModal(): Promise<HTMLElement> {
+    userEvent.click(screen.getByRole('button', { name: 'New folder' }));
+    const foundModal = await screen.findAllByRole('dialog');
+    expect(foundModal).toHaveLength(1);
+
+    return foundModal[0];
+  }
+
+  const dirs = makeDirectories();
+  const files = makeFiles();
+  const items = dirs.concat(files);
+
+  beforeEach(async () => {
+    const fileStorage = new MockFileStorage(dirs.concat(files));
+    makeContainer({ fileStorage });
+    const foundItems = await screen.findAllByRole('listitem');
+    expect(foundItems).toHaveLength(items.length);
+  });
+
+  it('selects directory correctly', async () => {
+    clickItem(0);
+    const NewItemModal = await getNewItemModal();
+
+    const foundButtons = await findAllByRole(NewItemModal, 'button');
+    const buttonContent = foundButtons.map(button => button.innerHTML);
+
+    expect(buttonContent).toContain('root');
+    expect(buttonContent).toContain('testdir0');
+    for (let i = 1; i < 5; i += 1) {
```

Comment:
dirs.length instead of 5
```suggestion
    for (let i = 1; i < dirs.length; i += 1) {
```